src/server/races/default-race.js

```diff
@@ -0,0 +1,279 @@
+
+		// Define default marble attributes
+		this.marbleAttributeList.color = function(attribute) {
+			const colorRegEx = /#(?:[0-9a-fA-F]{3}){1,2}$/g;
+			let match = attribute.match(colorRegEx);
+			return match === null ? undefined : match[0];
+		};
+		this.defaultMarbleAttributeList.color = function() {
+			return utility.randomHexColor();
+		};
+
+		this.marbleAttributeList.skinId = function(attribute, userId) {
+			let skinId = skins.idList[attribute];
+			if (typeof skinId === "undefined")
+				return undefined;
+
+			// Check if this user has permission to use this skin
+			if(	config.discord.enabled
+				&& skins.skinList[skinId].premium
+				&& !permissions.memberHasPermission(userId, "PREMIUM_SKINS")) {
+				return undefined;
+			}
+			return skinId;
+		};
+		this.defaultMarbleAttributeList.skinId = function() {
+			return "default";
+		};
+
+		// 2% chance for a entry's marbles to be of a random size between .3 and .6
+		this.defaultMarbleAttributeList.size = function() {
+			return Math.random() > .98 ? (.3 + Math.random() * .3) : 0.2;
+		};
+	}
+
+	//
+	// State-related functions
+	//
+
+	onStateEnter() {
+	}
+
+	// Returns a promise that resolves once preparation has finished, or null if this state can be skipped
+	onStatePreparing() {
+		return null;
+	}
+
+	onStateStarting() {
+		// Default point contribution rules
+		// Other race types may want to override this with their own
+		this.pointContributors = this.raceEntries.filter(entry => entry.contributesPoints).length;
+	}
+
+	onStateStarted() {
+		this.round.start = Date.now();
+		physicsWorld.openGates();
+	}
+
+	onStateFinished() {
+		this.finalizeRound();
+
+		physicsWorld.clearMarbles();
+		physicsWorld.closeGates();
+	}
+
+	//
+	// Non-state functions
+	//
+
+	// Returns true if the marble/entry limit has been reached
+	addRaceEntry(id, name, attributeString = "") {
+		// Don't add new player entries if the limit has been reached, or players that already entered
+		// Exception: Added bots with a undefined id (only spawnable by devs/operators)
+		if(typeof id === "string" && this.isRaceFull()
+		|| (typeof id === "string" && typeof this.raceEntries.find(playerEntry => id === playerEntry.id) !== "undefined") )
+			return this.isRaceFull();
+
+		let marbleAttributes = this.parseMarbleAttributes(id, attributeString);
+		let entry = new RaceEntry(id, name, marbleAttributes);
+		this.raceEntries.push(entry);
+
+		this.onNewRaceEntry(entry);
+
+		return this.isRaceFull();
+	}
+
+	// Returns a marbleAttributes object for the given user id and attributes string
+	parseMarbleAttributes(id, attributeString) {
+		let marbleAttributes = {};
+		let messageSections = attributeString.split(" ");
+		messageSections.length = Math.min(messageSections.length, 4); // At most, check only the first 4 words
+
+		for(let i = 0; i < messageSections.length; i++) {
+			Object.keys(this.marbleAttributeList).forEach( attribute => {
```

Comment:
Normally this would be the better option, but in the current case, `attribute` is a `string`, which we use to write to the correct properties for the `marbleAttributes` object. The given alternative gives us an `Object` instead, which wouldn't make this possible.

Reply:
But it doesn't..? Unless I am missing something or being a big dumb (which is a very real possibility), a `for ... in` loop gives you the object keys as strings, not the values?

Reply:
Apparently I stand corrected! I'm kind of baffled that iterating over an object's properties this way gives you `string`s instead of, well, the properties. Js remains unintuitive and weird as hell for me, it seems.

Reply:
Just for complete clarity;
The `in` keyword gives you the keys of the object:
```js
for (let key in object) { ... }
```

The `of` keyword gives you the values of the object:
```js
for (let value of object) { ... }
```